system/security/securesocket/securesocket.cpp

```diff
@@ -1142,7 +1142,147 @@ const char* strtok__(const char* s, const char* d, StringBuffer& tok)
+        return false;
+
+    if (!SSL_CTX_use_RSAPrivateKey(ctx, rsa))
+        return false;
+
+    return true;
+}
+
+static bool useCertificateChainPEMBuffer(SSL_CTX *ctx, const char *certbuf, int certlen=-1)
+{
+    // this routine based on code originally from:
+    // https://stackoverflow.com/questions/3810058/read-certificate-files-from-memory-instead-of-a-file-using-openssl
+
+    // can have multiple certs in buffer
+
+    BIO_ptr cbio(BIO_new_mem_buf(certbuf, certlen), ::BIO_free);
+    if (!cbio)
+        return false;
+
+    STACK_OF(X509_INFO) *infostk = PEM_X509_INFO_read_bio(cbio.get(), NULL, NULL, NULL);
+    if (!infostk)
+        return false;
+
+    bool first = true;
+    X509_INFO *infoval;
+    for (int i=0; i<sk_X509_INFO_num(infostk); i++)
+    {
+        infoval = sk_X509_INFO_value(infostk, i);
+        if (infoval->x509)
+        {
+            // First cert is server/main cert. Remaining, if any, are intermediate certs.
+            if (first)
+            {
+                first = false;
+
+                // Set server certificate. Note that this operation increments the
+                // reference count, which means that it is okay for cleanup to free it.
+                if (!SSL_CTX_use_certificate(ctx, infoval->x509))
+                {
+                    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+                    return false;
+                }
+
+                if (ERR_peek_last_error() != 0)
+                {
+                    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+                    return false;
+                }
+
+                // Get ready to store intermediate certs, if any.
+                SSL_CTX_clear_chain_certs(ctx);
+            }
+            else
+            {
+                // Add intermediate cert to chain.
+                if (!SSL_CTX_add0_chain_cert(ctx, infoval->x509))
+                {
+                    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+                    return false;
+                }
+
+                // Above function doesn't increment cert reference count. NULL the
+                // reference to it in order to prevent it from being freed during cleanup.
+                infoval->x509 = NULL;
+            }
+        }
+    }
+
+    sk_X509_INFO_pop_free(infostk, X509_INFO_free);
+    return true;
+}
+
+static bool loadVerifyLocationsPEMBuffer(SSL_CTX *ctx, const char *cacertbuf, int cacertlen=-1)
+{
+    // this routine based on code originally from:
+    // https://stackoverflow.com/questions/5052563/c-openssl-use-root-ca-from-buffer-rather-than-file-ssl-ctx-load-verify-locat
+
+    // can have multiple certs in buffer
+
+    BIO_ptr cbio(BIO_new_mem_buf(cacertbuf, cacertlen), ::BIO_free);
+    if (!cbio)
+        return false;
+
+    STACK_OF(X509_INFO) *infostk = PEM_X509_INFO_read_bio(cbio.get(), NULL, NULL, NULL);
+    if (!infostk)
+        return false;
+
+    X509_STORE *store = SSL_CTX_get_cert_store(ctx);
```

Comment:
I think this is leaked

Reply:
I am not sure.  I do not think we need to explicitly call X509_STORE_free(store), it _should_ be freed when the reference count of it goes to 0 ?

Reply:
right, yes, it's owned by the ctx, which will free it.